apps/opik-backend/src/main/java/com/comet/opik/domain/SpanService.java

```diff
@@ -244,4 +250,58 @@ public Mono<Boolean> validateSpanWorkspace(@NonNull String workspaceId, @NonNull
         return spanDAO.getSpanWorkspace(spanIds)
                 .map(spanWorkspace -> spanWorkspace.stream().allMatch(span -> workspaceId.equals(span.workspaceId())));
     }
+
+    @Trace(dispatcher = true)
+    public Mono<Void> create(@NonNull SpanBatch batch) {
+
+        if (batch.spans().isEmpty()) {
+            return Mono.empty();
+        }
+
+        List<String> projectNames = batch.spans()
+                .stream()
+                .map(Span::projectName)
+                .distinct()
+                .toList();
+
+        Mono<List<Span>> resolveProjects = Flux.fromIterable(projectNames)
+                .flatMap(this::resolveProject)
+                .collectList()
+                .map(projects -> bindSpanToProjectAndId(batch, projects))
+                .subscribeOn(Schedulers.boundedElastic());
+
+        return resolveProjects
+                .flatMap(spanDAO::batchInsert)
+                .then();
+    }
+
+    private List<Span> bindSpanToProjectAndId(SpanBatch batch, List<Project> projects) {
+        Map<String, Project> projectPerName = projects.stream()
+                .collect(Collectors.toMap(Project::name, Function.identity()));
+
+        return batch.spans()
+                .stream()
+                .map(span -> {
+                    String projectName = WorkspaceUtils.getProjectName(span.projectName());
+                    Project project = projectPerName.get(projectName);
+
+                    if (project == null) {
+                        throw new EntityAlreadyExistsException(new ErrorMessage(List.of("Project not found")));
+                    }
```

Comment:
This doesn't look like the appropriate exception for this. It'd be more some sort of 500 errors, because the project wasn't created in case of not found.

Reply:
Yeah, probaly I mistake, thanks for catching it